### PR TITLE
Make widgets really pluggable

### PR DESCRIPTION
**Pull request by Vlatombe**
`Widget` is historically an `ExtensionPoint`, however it is customizable only for views. This makes widgets customizable as well for `Job`s, `Computer` or `ComputerSet` (current widget users).

Introduce a `HasWidgets` interface holding generic facilities to look up applicable widgets, as well as binding widgets to relative urls based on their url names. The following classes are concerned
* `View`
* `Job`
* `ComputerSet`
* `Computer`
* `CloudSet`

Introduce a generic `WidgetFactory` to allow plugins to contribute widgets to the mentioned types.

This deprecates the pattern of annotating Widget implementations with `@Extension`, in favor of typed `WidgetFactory` annotated implementation.

* Moved views `ajaxExecutors` and `ajaxBuildQueue` back to the widgets they belong to. The specific logic is now in the `WidgetFactory` implementations (as `ajax` view).

* Add a `getUrl` method to Widget. This allows to resolve custom widget views under `widget/<widgetName>` both under the Jenkins root object, jobs and views.

* `<t:queue>` and `<t:executors>` taglibs usage outside of widget context is now deprecated. Instead, implement a `WidgetFactory` passing the relevant data context.

# Plugins impact

* Subclasses of `Widget` annotated with `@Extension` still works, but is deprecated. It is advised to remove `@Extension` annotation, implement a `WidgetFactory` and annotate it with `@Extension`.
* Implement `HasWidgets` when applicable and add relevant `WidgetFactory` subclasses as needed.
* Usages of Widgets removal should use an `ExtensionFilter` removing the corresponding `WidgetFactory` implementations.

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Testing done

* Create a pipeline job `p` as
```
node('some-missing-label') {} 
```
and launch it
* Main Jenkins view, check content of queue (1 item), check ajax requests (successful)
* Created custom empty view, check content of queue (1 item), check ajax requests (successful). Edit the view, tick Filter Build Queue and Filter build executors. Check that queue is now empty
* Browse `/computer/`, check widgets content and ajax requests
* Browse `/computer/(built-in)/`, check widgets content and ajax requests.
* Browse `/job/p`, check build history is displayed appropriately.

### Proposed changelog entries

- Developer: plugins can now contribute widgets for `Computer`, `ComputerSet`, `View`, `Job`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7932"></a>

### CODE REVIEW

File: core/src/main/resources/lib/hudson/queue.jelly
Comment (Vlatombe) [inline core/src/main/resources/lib/hudson/queue.jelly:106]: Breaking change, I need to assess how I can retain compatibility

File: core/src/main/resources/hudson/model/Computer/sidepanel.jelly
Comment (jglick) [inline core/src/main/resources/hudson/model/Computer/sidepanel.jelly:49]:
```suggestion
      <st:include it="${w}" page="index"/>
```

File: core/src/main/resources/hudson/model/ComputerSet/sidepanel.jelly
Comment (jglick) [inline core/src/main/resources/hudson/model/ComputerSet/sidepanel.jelly:40]:
```suggestion
      <st:include it="${w}" page="index"/>
```

File: core/src/main/resources/jenkins/widgets/BuildQueueWidget/ajax.jelly
Comment (jglick) [inline core/src/main/resources/jenkins/widgets/BuildQueueWidget/ajax.jelly:31]:
```suggestion
    <st:include page="index"/>
```

File: core/src/main/resources/jenkins/widgets/ExecutorsWidget/ajax.jelly
Comment (jglick) [inline core/src/main/resources/jenkins/widgets/ExecutorsWidget/ajax.jelly:31]:
```suggestion
    <st:include page="index"/>
```

**Review by jglick — COMMENTED** [review]
Concept looks good.

File: core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java:41]: All filtering is based on `isAssignableFrom` so this applies to any subtype as well.

File: core/src/main/java/jenkins/widgets/WidgetFactory.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/widgets/WidgetFactory.java:58]: That was copy-pasted from `TransientActionFactory`, forgot to remove it.

File: core/src/main/java/jenkins/widgets/WidgetFactory.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/widgets/WidgetFactory.java:100]: Probably wrong for `TransientActionFactory` as well.

File: core/src/main/resources/lib/hudson/executors.jelly
Comment (Vlatombe) [reply core/src/main/resources/lib/hudson/executors.jelly:162]: Boilerplace view like https://github.com/jenkinsci/cloudbees-folder-plugin/blob/a514313307249adcf9ee4f197af2d4f94dbd06f5/src/main/resources/com/cloudbees/hudson/plugins/folder/computed/FolderComputation/ajaxExecutors.jelly is now ignored.

File: core/src/main/resources/lib/hudson/queue.jelly
Comment (Vlatombe) [reply core/src/main/resources/lib/hudson/queue.jelly:106]: Boilerplate view like https://github.com/jenkinsci/cloudbees-folder-plugin/blob/a514313307249adcf9ee4f197af2d4f94dbd06f5/src/main/resources/com/cloudbees/hudson/plugins/folder/computed/FolderComputation/ajaxBuildQueue.jelly is now ignored.

File: core/src/main/resources/lib/hudson/executors.jelly
Comment (Vlatombe) [reply core/src/main/resources/lib/hudson/executors.jelly:162]: https://github.com/search?q=org%3Ajenkinsci%20ajaxExecutors&type=code shows no other occurrence in the community.

File: core/src/main/resources/lib/hudson/queue.jelly
Comment (Vlatombe) [reply core/src/main/resources/lib/hudson/queue.jelly:106]: More occurrences
https://github.com/search?q=org%3Ajenkinsci+ajaxBuildQueue&type=code
* simple-queue-plugin
* computer-queue-plugin
* maven-plugin
* ivy-plugin

Comment (Vlatombe) [reply core/src/main/resources/lib/hudson/queue.jelly:106]: Looked at the two first, doesn't seem to be a problem (they have their own widgets with their own views)

**Comment by Vlatombe** [comment]
Noting that https://github.com/jenkinsci/simple-queue-plugin/blob/7239902db60e4ae6c584bdbfe125d2652ad88fcd/src/main/java/cz/mendelu/xotradov/SimpleQueueInit.java#L24-L38 is broken but can easily be replaced by an `ExtensionFilter`.

File: core/src/main/java/jenkins/widgets/WidgetFactory.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/widgets/WidgetFactory.java]: Ended up removing the cache. Unlike actions, I don't think we're expecting more than 10 widgets implementations, and certainly not more than 2-3 per object. The actual lookup can still be revisited later with a cache for performance reasons.

Comment (jglick) [reply core/src/main/java/jenkins/widgets/WidgetFactory.java]: > Probably wrong for `TransientActionFactory` as well.

Possibly. I barely recall #2582 at this point.

**Comment by timja** [comment]
> For new APIs and extension points, there is a link to at least one consumer.

?

**Comment by Vlatombe** [comment]
> For new APIs and extension points, there is a link to at least one consumer.

Proprietary usage, I'm afraid.

File: core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java
Comment (jglick) [inline core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java:38]:
```suggestion
```

If it is not part of the API surface, then it does not matter when it was added.

File: core/src/main/resources/jenkins/agents/CloudSet/index.jelly
Comment (timja) [inline core/src/main/resources/jenkins/agents/CloudSet/index.jelly:30]: -1 page looks unbalanced, and off. see review comments in https://github.com/jenkinsci/cloud-stats-plugin/pull/73

**Comment by timja** [comment]
can you fill out testing done please.

File: core/src/main/java/hudson/widgets/BuildHistoryWidget.java
Comment (jglick) [inline core/src/main/java/hudson/widgets/BuildHistoryWidget.java:90]: Should probably be `@Restricted(DoNotUse.class)`?

File: core/src/main/java/hudson/widgets/Widget.java
Comment (jglick) [inline core/src/main/java/hudson/widgets/Widget.java:84]:
```suggestion
        String ownerUrl = getOwnerUrl();
        return (ownerUrl == null ? "" : ownerUrl) + "widget/" + getUrlName() + '/';
```
Would it make more sense for the default return value to be `""`?

File: core/src/main/java/jenkins/widgets/BuildQueueWidget.java
Comment (jglick) [inline core/src/main/java/jenkins/widgets/BuildQueueWidget.java:14]: Did you mean to delete all of this?

File: core/src/main/java/jenkins/widgets/ExecutorsWidget.java
Comment (jglick) [inline core/src/main/java/jenkins/widgets/ExecutorsWidget.java:80]: Need different symbols I think.

File: core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java
Comment (jglick) [inline core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java:55]: Should this be deprecated?

File: core/src/main/java/jenkins/widgets/WidgetFactory.java
Comment (jglick) [inline core/src/main/java/jenkins/widgets/WidgetFactory.java:87]: I think this could be restricted too?

Or both could just be inlined into `HasWidgets.getWidgets`.

**Review by jglick — APPROVED** [review]
Looks good. Minor suggestions/questions.

File: core/src/main/java/jenkins/widgets/ExecutorsWidget.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/widgets/ExecutorsWidget.java:80]: I don't think it really matters. A symbol must be unique within its extension point, but that's it.

Comment (jglick) [reply core/src/main/java/jenkins/widgets/ExecutorsWidget.java:80]: > A symbol must be unique within its extension point

Which in this case I would think is `WidgetFactory`, and you have three named `executors`.

Not sure there is any practical impact for now, since this is not something you can configure via JCasC. If https://plugins.jenkins.io/extension-filter/ were refreshed and made to support JCasC then it could be visible.

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/widgets/Widget.java
Comment (Vlatombe) [reply core/src/main/java/hudson/widgets/Widget.java:85]: Not sure I get the comment: default value if what?

**Comment by jglick** [comment]
https://github.com/jenkinsci/jenkins/pull/7932/checks?check_run_id=13810929549 seems like a flake from https://github.com/jenkinsci/jenkins-test-harness/pull/198 :eyes:

**Comment by Vlatombe** [comment]
@timja Any outstanding comment on this one? I think it is ready to go.

**Comment by timja** [comment]
> @timja Any outstanding comment on this one? I think it is ready to go.

none I haven't had time to review it fully but don't wait for me.

**Comment by Vlatombe** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by jglick — APPROVED** [review]

**Review by rsandell — APPROVED** [review]

File: core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java
Comment (basil) [inline core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java:40]: Is there a plan to enumerate consumers and adapt them to this deprecation?

Comment (basil) [reply core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java:40]: Should there be such a plan?

Comment (Vlatombe) [reply core/src/main/java/jenkins/widgets/JenkinsWidgetFactory.java:40]: I don't think it is necessary. Widgets contributing to Jenkins `Views` have no drawback staying on the old extension point, and implementing a `WidgetFactory` will allow contributions to new places.

File: core/src/main/java/hudson/model/Job.java
Comment (basil) [inline core/src/main/java/hudson/model/Job.java:618]: If this is deprecated for removal, then why is `AbstractProject#createHistoryWidget` still calling it?

**Comment by MarkEWaite** [comment]
This appears to have caused external job monitor plugin to no longer show build history.  [JENKINS-71553](https://issues.jenkins.io/browse/JENKINS-71553) includes more details

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [reply core/src/main/java/hudson/model/Job.java:618]: As is `WorkflowJob`. There is a reference to another method but it is unclear from the Javadoc how the deprecation is supposed to be resolved. By deleting the override?

Comment (Vlatombe) [reply core/src/main/java/hudson/model/Job.java:618]: All `Widget` instances are now instantiated via `WidgetFactory` extension point. Overrides should be deleted.

Filed https://github.com/jenkinsci/jenkins/pull/9268 to properly clean this up.

**Pull request merged**
